aws/src/main/java/org/apache/iceberg/aws/AwsClientFactories.java

```diff
@@ -88,6 +91,32 @@ private static AwsClientFactory loadClientFactory(String impl, Map<String, Strin
     return factory;
   }
 
+  /**
+   * Metrics are the only reliable way provided by the AWS SDK to determine if an API call was
+   * retried. Not for public use.
+   */
+  public static class InternalRetryDetector implements MetricPublisher {
```

Comment:
nit: for non-public classes, we should mark it as package protected or private

Reply:
actually given the fact that this is referred across class, seems like it should not be nested in AwsClientFactories? I get now why it's public. In that case we should name it something more public, maybe just `RetryDetector` is good enough.

Reply:
Will make it to its own utility class and rename it to `RetryDetector`